src/dyninst/test_thread_1_mutatee.c

```diff
@@ -1,177 +1,103 @@
- */
+/********************************************************************
+Subtest 1:  rtlib spinlocks
+*
+*  This tests the dyninst_lock_t implementation by modifying an
+*  atomic value while holding a lock.
+*
+********************************************************************/
 
-/* Global variables accessed by the mutator.  These must have globally unique
- * names.
- */
+static void (*DYNINSTinit_thelock)(dyninst_lock_t *);
+static void (*DYNINSTfree_thelock)(dyninst_lock_t *);
+static int (*DYNINSTlock_thelock)(dyninst_lock_t *);
+static void (*DYNINSTunlock_thelock)(dyninst_lock_t *);
 
-/* Internally used function prototypes.  These should be declared with the
- * keyword static so they don't interfere with other mutatees in the group.
- */
+static dyninst_lock_t test1lock;
+static Thread_t test1threads[TEST1_THREADS];
+static int subtest1err = 0;
+testsuite_atomic(pthread_t, canary, 0)
 
-/* Global variables used internally by the mutatee.  These should be declared
- * with the keyword static so they don't interfere with other mutatees in the
- * group.
+/*
+ * This barrier is a (maybe) temporary and incomplete fix for the threading
+ * issues in proccontrol's mailbox. It mitigates the bug there by ensuring
+ * that no thread in this mutatee can finish execution before all of the
+ * other threads are created.
  */
+static pthread_barrier_t startup_barrier;
 
-/* Function definitions follow */
-
-/********************************************************************/
-/********************************************************************/
-/***********  Subtest 1:  rtlib spinlocks */
-/***********  use dlopen/dlsym to get access to rt lib lock */
-/***********  then start up a bunch of threads to contend for it */
-/***********  monitor contention for deadlock/broken lock */
-/********************************************************************/
-/********************************************************************/
-
-unsigned long current_locks[TEST1_THREADS];
-/*Thread_t  *test2threads; */
-Thread_t test1threads[TEST1_THREADS];
-pthread_mutex_t real_lock;
-
-int subtest1counter = 0;
-int subtest1err = 0;
-
-void register_my_lock(unsigned long id, unsigned int val)
-{
-  unsigned int i;
-  int found = 0;
-  dprintf("%s[%d]:  %sregister lock for thread %lu\n", __FILE__, __LINE__,
-           val ? "" : "un", id);
-  for (i = 0; i < TEST1_THREADS; ++i) {
-    if (pthread_equal((pthread_t)test1threads[i],(pthread_t)id)) {
-      found = 1;
-      current_locks[i] = (unsigned)val;
-      break;
-    }
-  }
-  if (!found)
-    logerror("%s[%d]: FIXME\n", __FILE__, __LINE__);
-}
+static void *thread_main1(void *arg) {
+  pthread_barrier_wait(&startup_barrier);
 
-volatile int done_threads = 0;
+  // We need a unique value for each thread, so just use its pthread ID
+  pthread_t id = pthread_self();
 
-int all_threads_done()
-{
-  return done_threads == TEST1_THREADS;
-}
+  (*DYNINSTlock_thelock)(&test1lock);
```

Comment:
clearer as `DYNINSTlock_thelock(&test1lock)` and other places where the function pointer is used in a call.